src/array.cr

```diff
@@ -1045,6 +1026,64 @@ class Array(T)
     self
   end
 
+  # Inserts all of the elements from *other* before the element at *index*.
+  #
+  # This method shifts the element currently at *index* (if any) and any
+  # subsequent elements to the right, increasing their indices. If the value
+  # of *index* is negative, counting starts from the end of the array.
+  #
+  # Raises `IndexError` if the *index* is out of bounds.
+  #
+  # ```
+  # fruits = ["Apple"]
+  # newFruits = ["Dragonfruit", "Elderberry"]
+  #
+  # fruits.insert_all(1, newFruits)             # => ["Apple", "Dragonfruit", "Elderberry"]
+  # fruits.insert_all(-3, ["Banana", "Cherry"]) # => ["Apple", "Banana", "Cherry", "Dragonfruit", "Elderberry"]
+  #
+  # fruits.insert_all(6, ["invalid"])  # raises IndexError
+  # fruits.insert_all(-7, ["indices"]) # raises IndexError
+  # ```
+  def insert_all(index : Int, other : Enumerable(T) | Iterable(T)) : self
+    other_size = other.size
+
+    return self if other_size == 0
+
+    normalized_index = index < 0 ? index + @size + 1 : index
+
+    unless 0 <= normalized_index <= @size
+      raise IndexError.new "insertion index of #{index} should be between (0..#{@size}) or (-#{@size + 1}..-1)"
+    end
```

Comment:
**suggestion:** `Indexable#check_index_out_of_bounds` implements this already. It is used in other places to validated and normalize an index.